call MarkRemountRequired on terminating pods

DaGenix opened the pull request:
https://github.com/kubernetes/kubernetes/issues/116481 describes an issue where volumes stop being refreshed once a Pod enters the "Terminating" state. This is a problem for serviceAccount tokens being supplied via projected volumes. The net effect is that if a Pod is in the terminating state for more than 20% of the expiration time of the service account token, it may go stale. This is becoming a more significant issue with the change to limit Kubernetes service account tokens to 1 hour - since a terminationGracePeriod seconds of just 12 minutes could cause the token to become invalid. It's also a challenge for other serviceAccount use cases, such as AWS IAM Roles - where the max lifetime of the token of 1 day means that grace periods greater than 4.8 hours could result in an expired token.

In a running Pod, what happens is:

In one go-routine we:
1. We run findAndAddNewPods() a few times a second
2. That method calls processPodVolumes() (but only if the Pod isn't Terminating)
3. In processPodVolumes(), we check to see if podPreviouslyProcessed() returns true. If it does, we do nothing
4. Otherwise, we call MarkRemountRequired() which results in volumes being refreshed

In a second go-routine we:
1. Call SyncPod() every now and then.
2. SyncPod will call WaitForAttachAndMount()
3. WaitForAttachAndMount() calls ReprocessPod()
4. ReprocessPod() calls markPodProcessingFailed()
5. markPodProcessingFailed() causes the next call to podPreviouslyProcessed() to return false, which is what results in us refreshing volumes

In a terminating Pod, the first loop is still running. However, in the second loop we call SyncTerminatingPod() instead of SyncPod(). And SyncTerminatingPod() never calls WaitForAttachAndMount() which means that ultimately podPreviouslyProcessed() will return true until the Pod shuts down and we'll never refresh any volumes.

So, we have two problems:
1. findAndAddNewPods() skips all Pods that are Terminating. Which means we'll never refresh volumes for terminating Pods.
2. Even if we fix the above issue, when we actually get to processPodVolumes() the call to podPreviouslyProcessed() will always return true for a Terminating pod - so, we'll never refresh any volumes.

The approach to fix the issue in this PR is to change how podPreviouslyProcessed() works. Previously, it would return true up until markPodProcessingFailed() is called (as a result of SyncPod()). We update it so that it will return false if its been more than 60 seconds since markPodProcessed() has been called. This fixes the second problem above, since even though SyncTerminatingPod() won't call the methods we need it to, every 60 seconds processPodVolumes() will be allowed to run and refresh volumes.

The first problem is resolved by removing its check for if a Pod is terminating and then skipping it. Instead, we move that check into processPodVolumes() - where we check to see if the Pod is terminating, and, if so, all we do is refresh volumes and then bail out of the rest of the method were we might add in new volumes.

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
7. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
8. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
9. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
10. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind bug

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

Fixes an issue where projected volumes on terminating pods would not be refreshed. This is a problem, because security tokens on those pods could become stale while the pod is terminating.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #116481
Closes #117329

#### Special notes for your reviewer:

I'm a bit unclear if all tests will pass. However, I've validated that the change does appear to fix the problem. If this approach looks reasonable, I'm happy to make sure that the tests are passing.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Project volumes on terminating pods will continue to be refreshed until the Pod terminates.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
N/A

DaGenix commented:
NOTE: My previous attempt at resolving this issue was #117329, but there were some performance concerns with that approach.

bart0sh commented:
/triage accepted
/priority important-soon
/ok-to-test
/cc @gnufied @jsafrane

liggitt commented on pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:
```diff
@@ -197,6 +202,15 @@ func (dswp *desiredStateOfWorldPopulator) findAndAddNewPods() {
 			// Do not (re)add volumes for pods that can't also be starting containers
+			// However, we do need to potentially remount volumes while terminating -
+			// for example projected volumes
+			uniquePodName := util.GetUniquePodName(pod)
+			var entry = dswp.pods.processedPods[uniquePodName]
+			if time.Since(entry.lastMarkRemount) > time.Duration(10*time.Second) {
+				dswp.actualStateOfWorld.MarkRemountRequired(uniquePodName)
+				entry.lastMarkRemount = time.Now()
+				dswp.pods.processedPods[uniquePodName] = entry
+			}
```
making these changes here (marking remount required and directly manipulating processedPods) is _really_ hard to reason about and does not look safe
* markPodProcessed modifies processedPods under lock, this does not
* it is _really_ hard to understand the implications of marking this as remount required here, but then not calling processPodVolumes

I think we should keep the remount marking and processedPods update calls in the same place rather than introducing new places where we update those.

DaGenix replied:
Thanks for the feedback. I'm going to try to rework my changes to address your concerns.

liggitt reviewed (commented):
/hold

DaGenix commented:
@liggitt I pushed a new commit to try to address your feedback

liggitt commented on pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:
```diff
@@ -194,12 +199,6 @@ func (dswp *desiredStateOfWorldPopulator) findAndAddNewPods() {
 
 	for _, pod := range dswp.podManager.GetPods() {
-		// Keep consistency of adding pod during reconstruction
-		if dswp.hasAddedPods && dswp.podStateProvider.ShouldPodContainersBeTerminating(pod.UID) {
```
I don't know this subsystem in detail, but I doubt completely removing this is correct

The overall goal is to *keep* refreshing volumes *which have already been set up* as long as the containers they are mounted into are not yet terminated. I don't see those aspects considered here.

The thing we want to guard against is unnecessarily trying to set up *new* volumes for containers which have not yet started if the pod has gone into terminating state so those new containers will never be started. I don't see those aspects considered here either.

I think detailed guidance from sig-storage maintainers is needed to know how to accomplish both of the above goals simultaneously, and it's possible (likely, even) that the current volume setup / sync approach will need tweaking to allow that.

DaGenix replied:
Thanks for your feedback, @liggitt !

> The overall goal is to keep refreshing volumes which have already been set up as long as the containers they are mounted into are not yet terminated. I don't see those aspects considered here.

In this PR, I'm accounting for that by having `podPreviouslyProcessed()` start returning `false` if its been more than 60 seconds since `markPodProcessed()` was called. And this timeout works even if the Pod is in a terminating state and the loop that manages the Pod is stuck waiting for the Pod to terminate. Or, am I misunderstanding your comment?

> The thing we want to guard against is unnecessarily trying to set up new volumes for containers which have not yet started if the pod has gone into terminating state so those new containers will never be started. I don't see those aspects considered here either.

Ah, cool. I didn't take that into consideration. I will look into that.

DaGenix replied:
@liggitt I've pushed a new change that I hope addresses you feedback (assuming that I understand it correctly - apologies if I do not). I have also updated the PR description with a better summary of the change I'm proposing here. Thanks!

DaGenix replied:
@liggitt do you have any thoughts on my responses to your comments? Do you have any thoughts on how I could request guidance from the sig-storage folks? Thanks!

liggitt replied:
> I think detailed guidance from sig-storage maintainers is needed

I don't the expertise to review this, we *really* need storage subsystem experts to review this change closely

liggitt replied:
> In this PR, I'm accounting for that by having `podPreviouslyProcessed()` start returning `false` if its been more than 60 seconds since `markPodProcessed()` was called.

What is significant about 60 seconds? It is not clear to me how that scopes this change to exactly the containers and volumes we care about continuing to refresh

DaGenix replied:
Unless I'm misunderstanding, there are two loops currently being used:

1. The first, is that volume_manager.go creates a DesiredStateOfWorldPopulator [here](https://github.com/kubernetes/kubernetes/blob/e5ba5cd2b0d3bddd4a0393b98fde14960f31a33c/pkg/kubelet/volumemanager/volume_manager.go#L213). It starts it [here](https://github.com/kubernetes/kubernetes/blob/e5ba5cd2b0d3bddd4a0393b98fde14960f31a33c/pkg/kubelet/volumemanager/volume_manager.go#L294). It runs every [100 milliseconds](https://github.com/kubernetes/kubernetes/blob/e5ba5cd2b0d3bddd4a0393b98fde14960f31a33c/pkg/kubelet/volumemanager/volume_manager.go#L64). Its not really clear to me why its 100 milliseconds. Anyway, as best as I can tell, this loop mostly runs and does nothing until its triggered by the second loop.

2. The second loop is what I'm calling the SyncPod loop. Its defined [here](https://github.com/kubernetes/kubernetes/blob/e5ba5cd2b0d3bddd4a0393b98fde14960f31a33c/pkg/kubelet/kubelet.go#L2323). Once [every second](https://github.com/kubernetes/kubernetes/blob/e5ba5cd2b0d3bddd4a0393b98fde14960f31a33c/pkg/kubelet/kubelet.go#L2328), a timer fires, and through a chain of calls, that causes markPodProcessingFailed() to be called.

Anyway, the first loop does nothing most times it run. However, after markPodProcessingFailed() is called, that causes the first loop to jump into action and update the projected volumes, at least as I understand it.

So, why did I pick 60 seconds? I picked it because it seemed that a 60 second timeout was quite a bit longer than the existing loops (100ms and 1s) and so re-processing projected volumes every 60 seconds wouldn't be a performance issue when we were already handling them once a second. However, 60 seconds also seems short enough that any projected volumes that need to be updated will not have gotten stale in less than 60 seconds.

So, yes, its a bit of a magic number. But the other loops seem like they have magic numbers as well, unless there is something I'm missing - and if thats the case, please let me know.

matthyx replied:
@liggitt as I understand it you would be more willing to keep a refresh every 1s to match the syncPod periodicity during termination?

liggitt replied:
I don't know or understand the volume lifecycle loops well enough to give feedback on this approach, it just doesn't seem principled to me, and seems like it could easily start trying to refresh volumes in pods we don't want it to.

olyazavr commented:
Is this still being worked on? We've encountered problems that could be fixed with this and are interested in seeing this move forward

DaGenix commented:
@olyazavr - @liggitt gave me some feedback late last year, which I thought I addressed. What I need is for someone to let me know if my changes addressed the issues that were brought up / if my approach in this PR is correct.

DaGenix commented:
@mattcary @xing-yang @gnufied @jsafrane - I was curious if any of you had any feedback? I think I addressed @liggitt's feedback. Thanks!

DaGenix commented:
@mattcary @xing-yang @gnufied @jsafrane @liggitt - Is anyone able to provide me feedback? Should I check out the Slack channel? Thanks!

sftim commented:
Some tips on how to attract reviewers (who are often volunteers):

- I notice this PR isn't passing tests. If you can rebase it against _main_ and show that it passes all the tests, that gives a signal to reviewers that this is ready for more attention.
- This is a node PR. If you join the Zoom call for SIG Node, you can learn about the best way to get additional feedback on the code.
- Write the PR title and summary to start with a summary of the change, aimed at an audience that hardly has Kubernetes experience, but understands cloud IT pretty well.
- Read the Kubernetes guidance on writing changelog entries, and make sure you're following that fully.

SergeyKanzhelev commented:
This PR doesn't have any tests added. The change like this needs to be well tested. Please consider adding tests to e2e_node

matthyx commented:
@DaGenix I think your approach is good here, do you need any help reviving it for 1.32?

DaGenix commented:
@matthyx Thanks! I'll work on getting it rebased to the main branch and get the tests to pass. I see that merge commits are not allowed - so I will force-push after I rebase to fix that.

matthyx commented:
> @matthyx I rebased onto master. I also squashed everything down into a single commit with an updated commit message. All tests are passing. Please let me know if there is anything else that I should do. Thanks!

Very nice! I think @SergeyKanzhelev wanted a e2e test covering that.

liggitt commented:
/cc @msau42 @bobbypage @SergeyKanzhelev @smarterclayton 

for visibility to volume / pod lifecycle changes

DaGenix commented:
@liggitt very much appreciate any feedback that you or other experts can provide! I'm running low on other ideas for how this can be addressed outside of this approach. One idea I had, is that it appears to me that the main part of the problem is that when a pod enters the terminating state, we call SyncTerminatingPod instead of SyncPod. And the former method, unlike the latter, won't ping the DesiredStateOfWorldPopulator to refresh projected volumes - since SyncTerminatingPod waits synchronously for the pod to complete. One thing we could do, is to start up a separate go routine somewhere around this time, which would just ping the appropriate DesiredStateOfWorldPopulator method every few seconds to get it to reprocess projected volumes and then shut down this go routine after the pod actually terminates. I'm pretty sure that would work, but, it felt more surprising than the approach I've taken in this PR. But, I would be happy to try out that as an alternative it it made more sense. The second idea I had is to do a much more significant refactoring of how the pop lifecycles and volume handling works - but I fear that doing that would be well outside of the amount of time I can dedicate to this issue and my knowledge of how this code works.

@matthyx Thank you for your feedback! I do not currently know how to write an end-to-end test. I'd be happy to learn and do so, however! But, it feels like there is still some uncertainty that this approach is going to be accepted. I'd really like to feel with more certainty that this approach will be accepted before I spend some number of hours figuring out how to do that. Does that make sense? Thanks again!

DaGenix commented:
/remove-lifecycle rotten

DaGenix commented:
/remove-lifecycle stale

drjackild committed:
e2e: add test to cover SA token rotation during termination state

* as per discussion in #116481 add e2e test to check that token stops
  rotating when pod is terminating

DaGenix committed:
Refresh volumes on terminating pods

https://github.com/kubernetes/kubernetes/issues/116481 describes an
issue where volumes stop being refreshed once a Pod enters the
"Terminating" state. This is a problem for serviceAccount tokens being
supplied via projected volumes. The net effect is that if a Pod is in
the terminating state for more than 20% of the expiration time of the
service account token, it may go stale. This is becoming a more
significant issue with the change to limit Kubernetes service account
tokens to 1 hour - since a terminationGracePeriod seconds of just 12
minutes could cause the token to become invalid. It's also a challenge
for other serviceAccount use cases, such as AWS IAM Roles - where the
max lifetime of the token of 1 day means that grace periods greater than
4.8 hours could result in an expired token.

While a pod is running, SyncPod() will be called every now and then.
When it is called, it will then call WaitForAttachAndMount(). The result
of that call is that processPodVolumes(), when it runs, will call
MarkRemountRequired() for those pods. This results in volumes being
remounted. This is important for projected volumes since that is how
their contents are updated.

When a pod is in the terminating state, however, instead of SyncPod(),
SyncTerminatingPod() will run. And that function attempts to terminate
the pod and waits for that to complete - significantly without calling
WaitForAttachAndMount() at all - which means that projected volumes
won't be regenerated.

The approach that this commit takes to resolve this issue is to update
findAndAddNewPods(). Previusly, we'd ignore any terminating pods. Now,
we move the check that was ignoring terminating pods into
processPodVolumes(). And, instead of doing nothing for a terminating
pod, we now call MarkRemountRequired() in order to cause volumes to be
remounted.

The above change doesn't work by itself, however - the problem is that
processPodVolumes() bails out early if podPreviouslyProcessed() returns
true. Previoulsy, this value would be set to true once
processPodVolumes() did its thing - and so future calls would ignore the
pod. For a running, pod, however, the SyncPod() loop would occasionally
result in a call to markPodProcessingFailed() which would then result in
podPreviouslyProcessed() returning false again which would cause volumes
to be updated. As explained above, this doesn't work for terminating
pods. So, what we need is a way to get podPreviouslyProcessed() to
return false every now and then for terminating pods. In order to
achieve this, what we do is to change podPreviouslyProcessed() to keep
track of the last time that markPodProcessed() has been called for a
pod. If its been at least one minute since the last call, we start to
return false so as to cause volumes to be updated. Since this doesn't
depend on the SyncPod() loop, this will work even for terminating pods.

I don't believe this should cause any performance issues - in the
running pod case, the SyncPod() loop will result in the occasional
reprocessing of pods as it does today. And since we keep track of when
the last time we reprocessed a pod was and only reprocess it again if
its been at least 1 minute, this shouldn't involve significantly if any
additional pod reprocessing. In the terminating pod case, we will now
reprocess volumes once a minute - which is what we want to keep them up
to date.

DaGenix commented:
@drjackild - I added your commit

drjackild commented:
The failing tests on me, sorry for inconvenience. I'll take care of them.

drjackild committed:
Remove unused functions from e2e tests

drjackild committed:
Remove unused imports from test/e2e/auth/service_accounts.go

DaGenix commented:
@drjackild Thanks! Applied!

msau42 reviewed (commented):
Linking my comment from the previous attempt at the fix:
https://github.com/kubernetes/kubernetes/pull/117329#discussion_r1245576210

In general, we should restrict this change only for volume types that require periodic refreshing.

drjackild commented:
@DaGenix we discussed this issue with @msau42 and to limit the scope even further - it would be better to restrict volumes refresh only to projected volumes containing `serviceAccountToken`. We can always make our scope wider in the future, if needed (for example, to include CSI volumes with `csidriver.spec.republishrequired` enabled).

DaGenix commented:
/reopen

DaGenix commented:
/reopon

DaGenix commented:
/reopen

DaGenix commented:
/reopen

DaGenix commented:
/reopen

DaGenix commented:
/retest

DaGenix commented:
/reopen

DaGenix commented:
/retest

DaGenix commented:
/reopen

DaGenix commented:
/reopen

On Thu, Jul 16, 2026, 10:13 AM kubernetes-prow[bot] <
***@***.***> wrote:

> *kubernetes-prow[bot]* left a comment (kubernetes/kubernetes#122161)
> <https://github.com/kubernetes/kubernetes/pull/122161#issuecomment-4992902197>
>
> @k8s-triage-robot <https://github.com/k8s-triage-robot>: Closed this PR.
> Details
>
> In response to this
> <https://github.com/kubernetes/kubernetes/pull/122161#issuecomment-4992900922>
> :
>
> The Kubernetes project currently lacks enough active contributors to
> adequately respond to all issues and PRs.
>
> This bot triages PRs according to the following rules:
>
>    - After 90d of inactivity, lifecycle/stale is applied
>    - After 30d of inactivity since lifecycle/stale was applied,
>    lifecycle/rotten is applied
>    - After 30d of inactivity since lifecycle/rotten was applied, the PR
>    is closed
>
> You can:
>
>    - Reopen this PR with /reopen
>    - Mark this PR as fresh with /remove-lifecycle rotten
>    - Offer to help out with Issue Triage
>    <https://www.kubernetes.dev/docs/guide/issue-triage/>
>
> Please send feedback to sig-contributor-experience at kubernetes/community
> <https://github.com/kubernetes/community>.
>
> /close
>
> Instructions for interacting with me using PR comments are available here
> <https://git.k8s.io/community/contributors/guide/pull-requests.md>. If
> you have questions or suggestions related to my behavior, please file an
> issue against the kubernetes-sigs/prow
> <https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:>
> repository.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/122161?email_source=notifications&email_token=AAQA7W5TV4VFLR6CNKQQI7D5FDPHXA5CNFSNUABFM5UWIORPF5TWS5BNNB2WEL2JONZXKZKDN5WW2ZLOOQXTIOJZGI4TAMRRHE32M4TFMFZW63VHNVSW45DJN5XKKZLWMVXHJLDGN5XXIZLSL5RWY2LDNM#issuecomment-4992902197>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAQA7W6KU33U77WZRZWRW4D5FDPHXAVCNFSNUABEKJSXA33TNF2G64TZHMZDANJYGA2DSOB3JFZXG5LFHMZDAMRSGM3DOOJZGOQXMAQ>
> .
> Triage notifications, keep track of coding agent tasks and review pull
> requests on the go with GitHub Mobile for iOS
> <https://github.com/notifications/mobile/ios/AAQA7W3VT3RKF5N5NGG3WZ35FDPHXA5CNFSNUABFM5UWIORPF5TWS5BNNB2WEL2JONZXKZKDN5WW2ZLOOQXTIOJZGI4TAMRRHE32M4TFMFZW63VHNVSW45DJN5XKKZLWMVXHJKTGN5XXIZLSL5UW64Y>
> and Android
> <https://github.com/notifications/mobile/android/AAQA7W7TU4YF3NOL6TJZ4ET5FDPHXA5CNFSNUABFM5UWIORPF5TWS5BNNB2WEL2JONZXKZKDN5WW2ZLOOQXTIOJZGI4TAMRRHE32M4TFMFZW63VHNVSW45DJN5XKKZLWMVXHJLTGN5XXIZLSL5QW4ZDSN5UWI>.
> Download it today!
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

DaGenix commented:
/retest
